website/pages/docs/Cloud-Init/usage.mdx

```diff
@@ -0,0 +1,72 @@
+---
+layout: docs
+page_title: Vagrant Cloud-Init Usage
+sidebar_title: Usage
+description: Various Vagrant Cloud-Init examples
+---
+
+# Basic Usage
+
+~> **Warning!** This feature is experimental and may break or
+change in between releases. Use at your own risk. It currently is not officially
+supported or functional.
+
+This feature currently reqiures the experimental flag to be used. To explicitly enable this feature, you can set the experimental flag to:
+
+```
+VAGRANT_EXPERIMENTAL="cloud_init"
+```
+
+Please note that `VAGRANT_EXPERIMENTAL` is an environment variable. For more
+information about this flag visit the [Experimental docs page](/docs/experimental/)
+for more info. Without this flag enabled, any disks defined will not be configured.
+
+Also note that the examples below use the VirtualBox provider, which is the current
+supported providier for this feature.
+
+Below are some very simple examples of how to use Vagrant Disks with the VirtualBox provider.
+
+## Basic Examples
+
+Here is an example containing two parts of user data:
+
+```ruby
+config.vm.cloud_init do |cloud_init|
```

Comment:
So, although the RFC mentions having a method that is just called `add_user_config` that simply sets values for the cloud_init config class. I think this is an anti-pattern here. Current `config.vm` options such as `provision`, `synched_folder`, and `disk` configs are contained to a single ruby block, rather than multiple configs inside one ruby block. Not using a `add_user_config` here falls in line with the existing pattern Vagrant has defined for storing configuration. And having a config method/option called `add_user_config` seems redundant here, and makes the implementation for storing these config values more complicated than it needs to be if it's going to stay a `config.vm` option. Hopefully that makes sense for why I did that!

Reply:
Thanks for the explanation! The method should actually have been called `add_user_data`, which makes it a little less redundant. I saw this as somewhat akin to the Docker provisioner, which has a single configuration block that performs the same operation multiple times with different arguments:

```ruby
Vagrant.configure("2") do |config|
  config.vm.provision "docker" do |d|
    d.pull_images "ubuntu"
    d.pull_images "vagrant"
  end
end
```

The single block form also makes it possible to add other types of things to the cloud-init configuration in the future, like explicitly defining metadata:

```ruby
config.vm.cloud_init do |cloud_init|
  cloud_init.add_metadata instance_id: "i-95576fda42cf750a28"
end
```

In general, though, since we are producing a single cloud-init configuration, I think we should try to define that within the scope of a single config block.

I'm also sort of assuming that people are much less likely to use `inline` than `path`. What if we simplified the configuration to something like this?

```ruby
config.vm.cloud_init do |cloud_init|
  cloud_init.add_user_data "provision.sh", content_type: "text/x-shellscript"
  cloud_init.add_user_data "cloud.cfg", content_type: "text/cloud-config"
end
```

Reply:
> Thanks for the explanation! The method should actually have been called `add_user_data`, which makes it a little less redundant. I saw this as somewhat akin to the Docker provisioner, which has a single configuration block that performs the same operation multiple times with different arguments:
> 
> ```ruby
> Vagrant.configure("2") do |config|
>   config.vm.provision "docker" do |d|
>     d.pull_images "ubuntu"
>     d.pull_images "vagrant"
>   end
> end
> ```

Ah, sorry! I mistyped the name from the RFC. But I think my point still stands, we're already adding the data by defining it and storing it into the config, so I guess I just don't see it as a useful thing here and feels kind of redundant to me.

> 
> The single block form also makes it possible to add other types of things to the cloud-init configuration in the future, like explicitly defining metadata:
> 
> ```ruby
> config.vm.cloud_init do |cloud_init|
>   cloud_init.add_metadata instance_id: "i-95576fda42cf750a28"
> end
> ```

This could be done with a metadata hash that takes key=>values such as instance_id:

```ruby
config.vm.cloud_init do |cloud_init|
  cloud_init.metadata = {instance_id: "i-95576fda42cf750a28"}
end
```

> 
> In general, though, since we are producing a single cloud-init configuration, I think we should try to define that within the scope of a single config block.
> 
> I'm also sort of assuming that people are much less likely to use `inline` than `path`. What if we simplified the configuration to something like this?

I think it's better to be explicit when setting the path key rather than leaving it off.

> 
> ```ruby
> config.vm.cloud_init do |cloud_init|
>   cloud_init.add_user_data "provision.sh", content_type: "text/x-shellscript"
>   cloud_init.add_user_data "cloud.cfg", content_type: "text/cloud-config"
> end
> ```